### PR TITLE
refactor: generate ui schema for res def

### PR DESCRIPTION
**Pull request by thxCode**
**Problem:**

- keep the keys have the same name, but do not consider the type.
- have not shrunk the range, including description, default, enum, maximum, minimum, and so on.

**Solution:**

shrink the first-level properties of the schemas.

**Related Issue:**
#1603

### CODE REVIEW

File: pkg/apis/resourcedefinition/helper.go [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+				if period+1 < len(d) {
+					lv.Description = string(d[:period+1])
+				}
+			} else {
+				var rt string
+				if t := strings.ToLower(lv.Title); t != strs.Pluralize(t) {
+					rt = "Specify the " + t + "."
+				} else {
+					rt = "Specify " + t + "."
+				}
```
Comment (gitlawr) [inline pkg/apis/resourcedefinition/helper.go:291]: The logic depending the peroid seems fragile. What if the description uses other terminal punctuation like `?/!` or anthing else? Is there a more universal approach, like getting the longest common substring?

File: pkg/apis/resourcedefinition/helper.go [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+		if (lv.Default != nil && rv.Default == nil) ||
+			(lv.Default != nil && rv.Default != nil && !reflect.DeepEqual(lv.Default, rv.Default)) {
+			lv.Default = nil
+
+			setSchemaRefMutuallyExclusive(lsr, mef)
+		}
+	}
+
+	// Get the optimal range.
+	// If found exclusion case, it falls back to blank range.
```
Comment (gitlawr) [inline pkg/apis/resourcedefinition/helper.go:311]: The blank range seems invalid, can we remove the property in this case?

File: pkg/apis/resourcedefinition/helper.go [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+	"github.com/seal-io/walrus/utils/pointer"
+	"github.com/seal-io/walrus/utils/strs"
+)
+
+const (
+	variablesSchemaKey = "variables"
+	outputsSchemaKey   = "outputs"
+
+	originalExtensionKey          = "x-walrus-original"
+	mutuallyExclusiveExtensionKey = "x-walrus-mutually-exclusive"
```
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/apis/resourcedefinition/helper.go:31]: do not need to be aware of this, but we need a place to record the state during shrinking.

File: pkg/apis/resourcedefinition/helper.go (hunk 2 above)
Comment (thxCode) [reply pkg/apis/resourcedefinition/helper.go]: updated

File: pkg/apis/resourcedefinition/helper.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (thxCode) [reply pkg/apis/resourcedefinition/helper.go]: use a memory notebook to replace this extension key.

**Review by gitlawr — DISMISSED** [review]
LGTM

File: pkg/apis/resourcedefinition/helper.go [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+	"github.com/seal-io/walrus/pkg/templates/openapi"
+	"github.com/seal-io/walrus/utils/json"
+	"github.com/seal-io/walrus/utils/pointer"
+)
+
+const (
+	variablesSchemaKey = "variables"
+	outputsSchemaKey   = "outputs"
+
+	originalExtensionKey = "x-walrus-original"
```
Comment (aiwantaozi) [inline pkg/apis/resourcedefinition/helper.go:28]: Can we use the existed extension key?
https://github.com/seal-io/walrus/blob/6200ad68e4725668a3ab4c9f355cd95fe4b37f51/pkg/templates/openapi/ext.go#L89

File: pkg/apis/resourcedefinition/helper.go [hunk 5]
```diff
@@ -0,0 +1,506 @@
...
+	if lv.Format == "" && rv.Format != "" {
+		lv.Format = rv.Format
+	}
+
+	if !lv.WriteOnly && rv.WriteOnly {
+		lv.WriteOnly = rv.WriteOnly
+	}
+
+	// Find the most complete sentence.
+	// If not found, it cleans up the description.
```
Comment (aiwantaozi) [inline pkg/apis/resourcedefinition/helper.go:266]: Can these logics be refactored into smaller functions? like alignDescription, alignDefault, alignEnum..

Comment (thxCode) [reply pkg/apis/resourcedefinition/helper.go:266]: although we can do this, I wouldn't say I like a lot of little pieces that function everywhere.

the independent functions are usually used for reusing, but alignDescription/alignDefault/... only works in one place.

**Review by aiwantaozi — APPROVED** [review]

**Pull request merged**